crates/kubelet/src/device_plugin_manager/pod_devices.rs

```diff
@@ -0,0 +1,157 @@
+pub type ResourceAllocateInfo = HashMap<String, DeviceAllocateInfo>;
+/// Map of container device information, keyed by container name
+pub type ContainerDevices = HashMap<String, ResourceAllocateInfo>;
+
+/// PodDevices contains the map of Pods to allocated devices
+#[derive(Clone)]
+pub struct PodDevices {
+    /// Map of devices allocated to the Pod keyed by Pod UID
+    allocated_devices: Arc<Mutex<HashMap<String, ContainerDevices>>>,
+    client: kube::Client,
+}
+
+impl PodDevices {
+    pub fn new(client: kube::Client) -> Self {
+        PodDevices {
+            allocated_devices: Arc::new(Mutex::new(HashMap::new())),
+            client,
+        }
+    }
+
+    /// get_active_pods activePods is a method for listing active pods on the node so the
+    /// amount of device plugin resources requested by existing pods could be counted
+    /// when updating allocated devices
+    pub async fn get_active_pods(&self) -> anyhow::Result<HashSet<String>> {
+        // TODO: should this be namespaced?
+        let pod_client: Api<Pod> = Api::all(self.client.clone());
+        let pods = pod_client.list(&ListParams::default()).await?;
+        Ok(pods
+            .iter()
+            .map(|pod| {
+                pod.metadata
+                    .uid
+                    .clone()
+                    .expect("Pod uid should always be set but was not")
+            })
+            .collect::<HashSet<String>>())
+    }
+
+    /// get_pods returns the UIDs of all the Pods in the `PodDevices` map
+    pub fn get_pods(&self) -> HashSet<String> {
+        self.allocated_devices
+            .lock()
+            .unwrap()
+            .iter()
+            .map(|(p_uid, _container_map)| p_uid.clone())
+            .collect::<HashSet<String>>()
+    }
+
+    pub fn remove_pods(&self, pods_to_remove: HashSet<String>) -> anyhow::Result<()> {
+        let mut allocated_devices = self.allocated_devices.lock().unwrap();
+        pods_to_remove.iter().for_each(|p_uid| {
+            allocated_devices.remove(p_uid);
+        });
+        Ok(())
+    }
+
+    /// Returns the device Ids of all devices requested by a container
+    pub fn get_container_devices(
+        &self,
+        resource_name: &str,
+        pod_uid: &str,
+        container_name: &str,
+    ) -> Option<HashSet<String>> {
+        let allocated_devices = self.allocated_devices.lock().unwrap().clone();
+        if let Some(container_devices) = allocated_devices.get(pod_uid) {
+            if let Some(resource_devices) = container_devices.get(container_name) {
+                if let Some(device_allocate_info) = resource_devices.get(resource_name) {
+                    return Some(device_allocate_info.device_ids.clone());
+                }
+            }
+        }
+        None
+    }
+
+    pub fn get_allocated_devices(&self) -> DeviceIdMap {
+        let allocated_devices = self.allocated_devices.lock().unwrap().clone();
```

Comment:
I think you can get away without the clone here

Reply:
I am cloning it in order to grab the contents of the mutex and immediately release it, but i could hold it for the function instead if there seems to be a reason to